### ISSUE TITLE
update CMake testing prebuilts to 3.10

### ISSUE THREAD

**Issue by DanAlbert** [android/ndk#898]
(this is an internal cleanup bug and has nothing to do with updating the SDK's CMake; follow https://b.android.com to file a Studio bug if that's what you're looking for)

The SDK has been shipping 3.10 for a while, but the prebuilt CMake we use for testing the NDK is still on 3.6. We should update the prebuilt repo to 3.10.

**Comment by arturbac** [comment]
Would be nice to have cmake 3.11/3.12 in sdk, it changes a lot with using imported targets

**Comment by DanAlbert** [comment]
File a bug against studio. We don't ship CMake.

**Comment by arturbac** [comment]
I know that, but ndk ships android.toolchain.cmake and I tought that depends on ndk compatibility what cmake is in sdk available.

**Comment by DanAlbert** [comment]
I don't think we actually gain much by testing with 3.10. What we really need to be doing is testing with 3.6 (the oldest supported version) and something as new as possible. 3.10 isn't new enough to be interesting any more, so I'm going to push this until the SDK gets a newer version.

We should also start proactively testing with new versions of CMake as they enter RC.

**Comment by alexcohn** [comment]
What is the incentive of still testing with 3.6? I would say, <s>people</s> projects that stay with old CMake, will never dare to move to newer NDK.

**Comment by DanAlbert** [comment]
Same reason we still test Jelly Bean: it's the minimum supported version so it's likely to break. Testing earliest and latest is most likely to catch bugs, and we don't have the resources to test every version, so it's ba good compromise.

**Comment by alexcohn** [comment]
Well I wanted to ask, why keep 3.6 as supported CMake version in NDK 21?

**Comment by DanAlbert** [comment]
No real reason to drop it yet. So far a lot more effort has gone into supporting new versions than old versions. I don't know of any killer features in new versions.

**Comment by arturbac** [comment]
Oh,. that's not good are not aware of new CMake Features.
It's a pain for me to maitainin projects as i have still to support lower that 3.13/3.14 CMake for android.
The biggest change is about exported targets its include, library, sources, headers dependency chain.
I realy love (modern) CMake 3.14 concepts, as I only _once_ have to configure all dependencies on exported target then they are simply inherited, I can selectivly use of target modules like boost::filesystem instead of ${BOOST_LIBRARIES} etc.
CMake 3.13/3.14 simplifies live for project maintainers.

More Modern CMake - Deniz Bahadir - Meeting C++ 2018
https://www.youtube.com/watch?v=y7ndUhdQuU8

**Comment by alexcohn** [comment]
> Same reason we still test Jelly Bean: it's the minimum supported version

Minimum supported OS version is not same as minimum supported CMake version. Pushing the former means that developers who still care about JellyBean devices cannot upgrade to latest NDK. Pushing the latter means that developers who dare to upgrade to latest NDK may also need some changes in their build scripts because of CMake incompatibilities. But there enough breaking changes on such upgrade anyway: compiler, "stl", gradle, targeting new platform…

**Comment by alexcohn** [comment]
> It's a pain for me to maitainin projects as i have still to support lower than 3.13/3.14 CMake for android.

Is that because Android SDK ships with up to 3.10 only?

**Comment by arturbac** [comment]
alexcohn
1) Yes. I cannot force team to explicitly install cmake 3.14, part of team (android part) is forced only to updating sdk and ndk
2) CMake 3.14 is backward compatibile, it can handle older cmake projects, sometimes You are only warned to explicilty specify some CMake policy to OLD or NEW.

**Comment by arturbac** [comment]
Example from one of projects used on linux (webservices, tools, desktop applications) and on android as application - library

if(ANDROID)
  target_link_libraries ( XXXX
    ${ZLIB_LIBRARIES}
    ${Boost_LIBRARIES}
    ${PLATFORM_CORE_LIBRARIES}
    )
else()
  target_link_libraries ( XXXX
    YYY
    ZLIB::ZLIB
    Boost::filesystem 
    Boost::system 
    Boost::thread 
    )
endif()

**Comment by DanAlbert** [comment]
> It's a pain for me to maitainin projects as i have still to support lower than 3.13/3.14 CMake for android.

You can do whatever you like with _your_ minimum supported version. If you want to tell your users that they have to use 3.14 that's your decision.

> The biggest change is about exported targets its include, library, sources, headers dependency chain.
> I realy love (modern) CMake 3.14 concepts, as I only once have to configure all dependencies on exported target then they are simply inherited, I can selectivly use of target modules like boost::filesystem instead of ${BOOST_LIBRARIES} etc.

These features are available in 3.6.

> Pushing the latter means that developers who dare to upgrade to latest NDK may also need some changes in their build scripts because of CMake incompatibilities.

They also need to update CMake. The CMake version available from various distros' package managers is often is quite old.

**Comment by DanAlbert** [comment]
@arturbac that non-Android side of that example works just fine on CMake 3.6. All of the CMake target properties stuff was a 3.0 feature.

**Comment by arturbac** [comment]
@DanAlbert - 3.6 its doesn't fully supports all features on setting private, public etc attributes with easy and clean syntax. You need for that 3.12, this is the lowest version with all capabilities. Watch the YT video from conference and You will understand that 3.6 to 3.12 is like C++98 to C++11.
I tried with 3.10 but it simply doesn't work in modern way, i need a lot of unclean coding with setting properties instead of using target_xxxx.
For example on linux requiring cmake 3.14 it will be very problematic if it is not available in packages for user linux distribution and missing in sdk. There a lot of possible problems for software that is not included in SDK or NDK on linux and windows.

**Comment by DanAlbert** [comment]
As I said, you're welcome to use CMake 3.14. That's not what this bug is about.

> On linux requiring cmake 3.14 it will be very problematic if it is not available in packages for user linux distribution and missing in sdk. There a lot of possible problems for software that is not included in SDK or NDK on linux and windows.

That has nothing to do with what's being discussed here. We're talking about the minimum version of CMake that we maintain compatibility for with the CMake toolchain file included in the NDK. We do not ship CMake at all.

**Comment by DanAlbert** [comment]
The NDK is testing with 3.16 now thanks to @hhb: https://android-review.googlesource.com/c/platform/prebuilts/cmake/linux-x86/+/1247088.

To reiterate: this is unrelated to what is available in the SDK manager. That work is separate but is being pursued.
